src/H5FDint.c

```diff
@@ -2392,6 +2434,431 @@ H5FD_sort_vector_io_req(hbool_t *vector_was_sorted, uint32_t _count, H5FD_mem_t
+            if ((fixed_element_sizes_index == count) && (element_sizes[i] == 0))
+                fixed_element_sizes_index = i - 1;
+            if ((fixed_bufs_index == count) && (bufs[i].cvp == NULL))
+                fixed_bufs_index = i - 1;
+        }
+
+        HDassert(fixed_element_sizes_index <= count);
+        HDassert(fixed_bufs_index <= count);
+
+        /* Populate the sorted arrays.  Note that the index stored in srt_tmp
+         * refers to the index in the unsorted array, while the position of
+         * srt_tmp within the sorted array is the index in the sorted arrays */
+        for (i = 0; i < count; i++) {
+
+            j = srt_tmp[i].index;
+
+            (*s_mem_space_ids_ptr)[i]  = mem_space_ids[j];
+            (*s_file_space_ids_ptr)[i] = file_space_ids[j];
+            (*s_offsets_ptr)[i]        = offsets[j];
+            (*s_element_sizes_ptr)[i]  = element_sizes[MIN(j, fixed_element_sizes_index)];
+            (*s_bufs_ptr)[i]           = bufs[MIN(j, fixed_bufs_index)];
+        }
+    }
+
+done:
+    if (srt_tmp) {
+        HDfree(srt_tmp);
+        srt_tmp = NULL;
+    }
+
+    /* On failure, free the sorted arrays if they were allocated.
+     * Note that we only allocate these arrays if the original array
+     * was not sorted -- thus we check both for failure, and for
+     * the flag indicating that the original array was not sorted
+     * in increasing address order.
+     */
+    if ((ret_value != SUCCEED) && (!(*selection_was_sorted))) {
+
+        /* free space allocated for sorted arrays */
+        if (*s_mem_space_ids_ptr) {
+            HDfree(*s_mem_space_ids_ptr);
+            *s_mem_space_ids_ptr = NULL;
+        }
+
+        if (*s_file_space_ids_ptr) {
+            HDfree(*s_file_space_ids_ptr);
+            *s_file_space_ids_ptr = NULL;
+        }
+
+        if (*s_offsets_ptr) {
+            HDfree(*s_offsets_ptr);
+            *s_offsets_ptr = NULL;
+        }
+
+        if (*s_element_sizes_ptr) {
+            HDfree(*s_element_sizes_ptr);
+            *s_element_sizes_ptr = NULL;
+        }
+
+        if (*s_bufs_ptr) {
+            HDfree(*s_bufs_ptr);
+            *s_bufs_ptr = NULL;
+        }
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+
+} /* end H5FD_sort_selection_io_req() */
+
+#ifdef H5_HAVE_PARALLEL
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD_selection_build_types
+ *
+ * Purpose:     ??
+ *              (derived from H5D__link_piece_collective_io() in src/H5Dmpio.c)
+ *              (also reference H5FD__mpio_vector_build_types() in src/H5FDmpio.c)
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ * Programmer:
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5FD_selection_build_types(hbool_t io_op_write, uint32_t num_pieces, H5S_t **file_spaces, H5S_t **mem_spaces,
```

Comment:
After further thought, I think we should move this function to H5FDmpio.c, and return this functionality to being integrated into link_piece_collective_io (with the frees moved to before the I/O as suggested in my email). I think it'd also be a good idea to move the check for num_pieces out of the function (and into the read/write callbacks), and remove the size_i  and mpi_buds_base parameters (calculate them in the read/write_callbacks).

Reply:
I will move this function to H5FDmpio.c as indicated.
And I will restore link_piece_collective_io() back to its original state in H5Dmpio.c (with the frees moved to before the IO), is that correct?

Reply:
That's correct, thanks!

Reply:
Changes made as indicated for H5FD_selection_build_types().
However, I still need to pass num_pieces to the routine because it is needed to malloc the arrays of mtype, ftype and displacement etc.  I can calculate mpi_bufs_base in the read/write callbacks, but I still need to pass the calculated info to H5FD_selection_build_types() for calculating displacements of the pieces.